REF: dataframe formatters/outputs

ivanovmg opened the pull request:
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Partially addresses #36407

This is a continuation of https://github.com/pandas-dev/pandas/pull/36434.

- Separated ``StringFormatter`` (or better ``ConsoleFormatter``?) from ``DataFrameFormatter``. Placed it into the new module (subject to discussion).
- Used composition of ``DataFrameFormatter`` in ``HTMLFormatter``, ``LatexFormatter``, ``CSVFormatter`` and ``StringFormatter``. It turned out that the inheritance of each of the formatters from the base ``DataFrameFormatter`` was too complicated to comprehend. The composition seems to suit here better.
- Created new class ``DataFrameRenderer`` to keep methods for outputs in each of the formats.

This is not the ultimate refactoring, but just one more step.

ivanovmg committed:
REF: drop TableFormatter

Move all methods to DataFrameFormatter,
inherit relevant classes from DataFrameFormatter.

ivanovmg committed:
REF: extract ConsoleFormatter

ivanovmg committed:
CLN: remove ConsoleFormatter to_string method

Replace it with get_result method, which is going to become
abstract method for the parent class.

ivanovmg committed:
REF: move table_id & render_links to HTMLFormatter

ivanovmg committed:
REF: move _join_multiline to ConsoleFormatter

ivanovmg committed:
REF: separate dataframe formatting from rendering

ivanovmg committed:
REF: extract _empty_info_line property in latex

ivanovmg committed:
DOC: docstrings for DataFrame & String Formatters

ivanovmg committed:
REF: make _get_buffer private

ivanovmg committed:
REF: pass DataFrameFormatter to CSVFormatter

ivanovmg committed:
REF: create to_csv in DataFrameRenderer

ivanovmg committed:
LINT: imports and line breaks

ivanovmg committed:
REF: move StringFormatter to separate module

New module suggested: pandas/io/formats/string.py

ivanovmg committed:
TYP: handle mypy errors after enabling composition

ivanovmg committed:
REF: remove non-existent parent in LatexFormatter

ivanovmg committed:
REF: move line_width to StringFormatter

ivanovmg committed:
REF: move need_to_wrap to StringFormatter

ivanovmg committed:
DOC: add docstrings to DataFrame.to_xxx methods

ivanovmg committed:
CLN: to_string on top in HTMLFormatter

ivanovmg committed:
LINT: black conv multiline -> oneline

jreback commented on pandas/io/formats/format.py:
```diff
@@ -1198,6 +899,209 @@ def _get_column_name_list(self) -> List[str]:
+
+        return None
+
+    def _get_result(
```
these can probably be non-private. This entire class is private, so that if these are called by other modules (from pandas) then we want to make the public. If something is entirely private to this class then this is ok.

ivanovmg replied:
It turns out that ``_get_result`` and ``_get_buffer`` should not belong to the class: they do not use the object state. So I decided to move the methods to the module level.

jreback commented on pandas/io/formats/format.py:
```diff
@@ -1198,6 +899,209 @@ def _get_column_name_list(self) -> List[str]:
+            raise TypeError("buf is not a file name and it has no write method")
+
+
 # ----------------------------------------------------------------------
```
would not be averse to splitting this file up if ppossible (e.g. maybe make array.py) for all of the formatters.

jreback replied:
certainly better as a followon

jreback commented on pandas/io/formats/format.py:
```diff
@@ -1198,6 +899,209 @@ def _get_column_name_list(self) -> List[str]:
         return names
 
 
+class DataFrameRenderer:
```
can you add a doc-string here and explain that these to_* are being called in core/frame/to_* ultimately.

ivanovmg replied:
Added

jreback reviewed (changes requested):
a few comments, looks good. nice reorg and cleanup. it has been a long time / never since this was done; and we have *lots* of built up code here.

ivanovmg committed:
DOC: add docstring to DataFrameRenderer

ivanovmg committed:
REF: move _get_result, _get_buffer to module level

These methods do not depend on the object state,
so were removed from the class to the module level.

ivanovmg committed:
REF: replace setters with initializer methods

ivanovmg committed:
REF: extract properties to make composition clear

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -2467,29 +2466,29 @@ def to_html(
-            table_id=table_id,
-            render_links=render_links,
         )
         # TODO: a generic formatter wld b in DataFrameFormatter
```
is this TODO still needed?

ivanovmg replied:
I don't know what it actually means here... :)

jbrockmendel commented:
there's a lot going on here.  can you mark which parts are pure-cut/paste and which are not?

ivanovmg commented on pandas/io/formats/string.py:
```diff
@@ -0,0 +1,200 @@
+"""
+Module for formatting output data in console (to string).
+"""
```
The content of this module and class StringFormatter was extracted from DataFrameFormatter.
Without substantial refactor only the parts related to console/string formatting were moved out of DataFrameFormatter.

ivanovmg commented on pandas/io/formats/format.py:
```diff
@@ -450,95 +455,8 @@ def get_adjustment() -> TextAdjustment:
-    columns: Index
-    _is_truncated: bool
-
-    @property
-    def is_truncated(self) -> bool:
-        return self._is_truncated
-
-    @property
-    def should_show_dimensions(self) -> bool:
-        return self.show_dimensions is True or (
-            self.show_dimensions == "truncate" and self.is_truncated
-        )
```
Moved to DataFrameFormatter

ivanovmg commented on pandas/io/formats/format.py:
```diff
@@ -833,223 +783,7 @@ def _get_strcols_without_index(self) -> List[List[str]]:
-    ) -> Optional[str]:
-        """
-        Render a DataFrame to a LaTeX tabular/longtable environment output.
-        """
```
Moved to DataFrameRenderer after some refactor.

ivanovmg commented on pandas/io/formats/format.py:
```diff
@@ -1062,34 +796,17 @@ def _format_col(self, i: int) -> List[str]:
             leading_space=self.index,
         )
 
-    def to_html(
-        self,
-        buf: Optional[FilePathOrBuffer[str]] = None,
-        encoding: Optional[str] = None,
-        classes: Optional[Union[str, List, Tuple]] = None,
-        notebook: bool = False,
-        border: Optional[int] = None,
-    ) -> Optional[str]:
```
Moved to DataFrameRenderer after some refactor.

ivanovmg commented on pandas/io/formats/format.py:
```diff
@@ -1062,34 +796,17 @@ def _format_col(self, i: int) -> List[str]:
-        return Klass(self, classes=classes, border=border).get_result(
-            buf=buf, encoding=encoding
-        )
+    def _get_formatter(self, i: Union[str, int]) -> Optional[Callable]:
+        if isinstance(self.formatters, (list, tuple)):
+            if is_integer(i):
+                i = cast(int, i)
+                return self.formatters[i]
+            else:
+                return None
+        else:
+            if is_integer(i) and i not in self.columns:
+                i = self.columns[i]
+            return self.formatters.get(i, None)
```
Moved from TableFormatter (which is deleted)

ivanovmg commented on pandas/io/formats/format.py:
```diff
@@ -1132,22 +849,6 @@ def space_format(x, y):
         # self.str_columns = str_columns
         return str_columns
 
-    @property
-    def has_index_names(self) -> bool:
-        return _has_names(self.frame.index)
-
-    @property
-    def has_column_names(self) -> bool:
-        return _has_names(self.frame.columns)
-
-    @property
-    def show_row_idx_names(self) -> bool:
-        return all((self.has_index_names, self.index, self.show_index_names))
-
-    @property
-    def show_col_idx_names(self) -> bool:
-        return all((self.has_column_names, self.show_index_names, self.header))
-
```
Re-ordered methods

ivanovmg commented on pandas/io/formats/format.py:
```diff
@@ -2023,26 +1942,6 @@ def set_eng_float_format(accuracy: int = 3, use_eng_prefix: bool = False) -> Non
     set_option("display.column_space", max(12, accuracy + 9))
 
 
-def _binify(cols: List[int], line_width: int) -> List[int]:
-    adjoin_width = 1
-    bins = []
-    curr_width = 0
-    i_last_column = len(cols) - 1
-    for i, w in enumerate(cols):
-        w_adjoined = w + adjoin_width
-        curr_width += w_adjoined
-        if i_last_column == i:
-            wrap = curr_width + 1 > line_width and i > 0
-        else:
-            wrap = curr_width + 2 > line_width and i > 0
-        if wrap:
-            bins.append(i)
-            curr_width = w_adjoined
-
-    bins.append(len(cols))
-    return bins
-
-
```
Moved close to StringFormatter

ivanovmg commented on pandas/io/formats/format.py:
```diff
@@ -833,223 +783,7 @@ def _get_strcols_without_index(self) -> List[List[str]]:
 
         return strcols
 
-    def _get_strcols(self) -> List[List[str]]:
-        strcols = self._get_strcols_without_index()
-
-        str_index = self._get_formatted_index(self.tr_frame)
-        if self.index:
-            strcols.insert(0, str_index)
-
-        return strcols
-
-    def _to_str_columns(self) -> List[List[str]]:
-        """
-        Render a DataFrame to a list of columns (as lists of strings).
-        """
-        strcols = self._get_strcols()
-
-        if self.is_truncated:
-            strcols = self._insert_dot_separators(strcols)
-
-        return strcols
```
Re-ordered/renamed

ivanovmg commented on pandas/io/formats/format.py:
```diff
@@ -833,223 +783,7 @@ def _get_strcols_without_index(self) -> List[List[str]]:
-    def _dimensions_info(self) -> str:
-        return f"\n\n[{len(self.frame)} rows x {len(self.frame.columns)} columns]"
-
-    def _get_string_representation(self) -> str:
-        if self.frame.empty:
-            info_line = (
-                f"Empty {type(self.frame).__name__}\n"
-                f"Columns: {pprint_thing(self.frame.columns)}\n"
-                f"Index: {pprint_thing(self.frame.index)}"
-            )
-            return info_line
-
-        strcols = self._to_str_columns()
-
-        if self.line_width is None:
-            # no need to wrap around just print the whole frame
-            return self.adj.adjoin(1, *strcols)
-
-        if self.max_cols is None or self.max_cols > 0:
-            # need to wrap around
-            return self._join_multiline(*strcols)
-
-        # max_cols == 0. Try to fit frame to terminal
-        return self._fit_strcols_to_terminal_width(strcols)
-
-    def _fit_strcols_to_terminal_width(self, strcols) -> str:
-        from pandas import Series
```
To StringFormatter

ivanovmg commented:
> 
> 
> there's a lot going on here. can you mark which parts are pure-cut/paste and which are not?

Did that partially. I hope it helps.

ivanovmg committed:
REF: eliminate inheritance for HTMLFormatter

ivanovmg committed:
LINT: new black

ivanovmg commented on pandas/io/formats/csvs.py:
```diff
@@ -79,11 +77,11 @@ def __init__(
-        self.decimal = decimal
-        self.header = header
-        self.index = index
+        self.na_rep = self.fmt.na_rep
```
Refactored to properties.

jreback commented on pandas/io/formats/csvs.py:
```diff
@@ -79,35 +77,45 @@ def __init__(
+        self.chunksize = self._initialize_chunksize(chunksize)
+
+    @property
+    def na_rep(self):
```
ideally if you can type these

jreback commented on pandas/io/formats/format.py:
```diff
@@ -1198,6 +899,224 @@ def _get_column_name_list(self) -> List[str]:
+        )
+        csv_formatter.save()
+
+        if path_or_buf is None:
```
would be nice to use save_to_buffer here as welll (in the future)

jreback commented on pandas/io/formats/string.py:
```diff
@@ -0,0 +1,200 @@
+            col.insert(row_num + n_header_rows, dot_str)
+        return strcols
+
+    def _join_multiline(self, *args) -> str:
```
ideally type these args

jreback commented on pandas/io/formats/string.py:
```diff
@@ -0,0 +1,200 @@
+            start = end
+        return "\n\n".join(str_lst)
+
+    def _fit_strcols_to_terminal_width(self, strcols) -> str:
```
ideally type strcols

jreback reviewed (changes requested):
couple of small comments, also pls rebase as merged the test changes.

ivanovmg committed:
Merge branch 'master' into refactor/fmt

ivanovmg committed:
TYP: type properties in CSVFormatter

ivanovmg committed:
TYP: type strcols

ivanovmg committed:
TYP: _join_multiline

ivanovmg committed:
LINT: sort in one line

ivanovmg committed:
REF: _join_multiline to accept single arg

simonjayhawkins commented on pandas/io/formats/string.py:
```diff
@@ -0,0 +1,199 @@
+        lwidth = self.line_width
+        adjoin_width = 1
+        if self.fmt.index:
+            idx = strcols.pop(0)
```
AFAICT the original _join_multiline didn't mutate the input.

often better to type inputs as Iterable instead of List for non-mutating functions

ivanovmg replied:
Oh, you are right! It did not mutate strcols, but now it does. I will fix it.

What I don't like about Iterable type is the following:
- String is Iterable, so it can be confused with List/Tuple.
- Cannot find len(iterable) - mypy throws
```Argument 1 to "len" has incompatible type "Iterable[str]"; expected "Sized"  [arg-type]```

So, now I changed the type of input parameter in ``_join_multiline`` to ``Iterable[List[str]]``.

simonjayhawkins replied:
> * String is Iterable, so it can be confused with List/Tuple.

yeah that can be a problem, but we normally add type parameters, so only Iterable[str] becomes an issue

> * Cannot find len(iterable) - mypy throws

can always use Sequence if needed as Sequence is also not mutable.

There are some guidelines for stubs at https://github.com/python/typeshed/blob/master/CONTRIBUTING.md#stub-file-coding-style

> avoid invariant collection types (List, Dict) in argument positions, in favor of covariant types like Mapping or Sequence

at some point, the guidelines there relevant to the codebase and not just stubs should probably be added to our code style docs #33851

ivanovmg committed:
REF: eliminate mutation in _join_multiline

jreback reviewed (approved):
lgtm. @jbrockmendel @simonjayhawkins if any comments.

simonjayhawkins commented:
don't wait for me one this. prioritising reviewing 1.1.x issues and PRs

ivanovmg committed:
Merge branch 'master' into refactor/fmt

ivanovmg committed:
Merge branch 'master' into refactor/fmt

ivanovmg commented:
Internal error in Windows py38_np18 check.
Some weird error in travis. Looks like internal travis error with closing buffer. Locally this test runs fine.
<details>
```
____________ test_filepath_or_buffer_arg[pathlike-foo-abc-to_latex] ____________

[gw0] linux -- Python 3.7.8 /home/travis/miniconda3/envs/pandas-dev/bin/python

method = 'to_latex'

filepath_or_buffer = PosixPath('/tmp/pytest-of-travis/pytest-0/popen-gw0/test_filepath_or_buffer_arg_pa11/foo')

assert_filepath_or_buffer_equals = <function assert_filepath_or_buffer_equals.<locals>._assert_filepath_or_buffer_equals at 0x7f368ef02170>

encoding = 'foo', data = 'abc', filepath_or_buffer_id = 'pathlike'

    @pytest.mark.parametrize("method", ["to_string", "to_html", "to_latex"])

    @pytest.mark.parametrize(

        "encoding, data",

        [(None, "abc"), ("utf-8", "abc"), ("gbk", "造成输出中文显示乱码"), ("foo", "abc")],

    )

    def test_filepath_or_buffer_arg(

        method,

        filepath_or_buffer,

        assert_filepath_or_buffer_equals,

        encoding,

        data,

        filepath_or_buffer_id,

    ):

        df = DataFrame([data])

    

        if filepath_or_buffer_id not in ["string", "pathlike"] and encoding is not None:

            with pytest.raises(

                ValueError, match="buf is not a file name and encoding is specified."

            ):

                getattr(df, method)(buf=filepath_or_buffer, encoding=encoding)

        elif encoding == "foo":

            with tm.assert_produces_warning(None):

                with pytest.raises(LookupError, match="unknown encoding"):

>                   getattr(df, method)(buf=filepath_or_buffer, encoding=encoding)

pandas/tests/io/formats/test_format.py:3389: 

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <contextlib._GeneratorContextManager object at 0x7f368ef0fcd0>

type = None, value = None, traceback = None

    def __exit__(self, type, value, traceback):

        if type is None:

            try:

>               next(self.gen)

E               AssertionError: Caused unexpected warning(s): [('ResourceWarning', ResourceWarning("unclosed <ssl.SSLSocket fd=13, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('10.20.0.43', 49154), raddr=('52.217.93.220', 443)>"), '/home/travis/build/pandas-dev/pandas/pandas/core/generic.py', 5404)]
```
</details>

simonjayhawkins reviewed (approved):
Thanks @ivanovmg generally lgtm.

The removal of the setters in csvs is a great improvement, for readability and static typing.

I'm not sure about removing the TableFormatter class though. This provided a common base for LatexFormatter, HTMLFormatter and what is now StringFormatter, for code that does not belong in DataFrameFormatter.

I suggest that future refactors are more atomic. I see several themes here that probably should have been discussed/reviewed in isolation. The primary one being the adoption of DataFrameFormatter for use in csv formatting. I'm also not sure that DataFrameRenderer class is even necessary.

simonjayhawkins commented:
> Internal error in Windows py38_np18 check.
> Some weird error in travis. Looks like internal travis error with closing buffer. Locally this test runs fine.

I've restarted the jobs. probably unrelated.

ivanovmg commented:
> 
> 
> Thanks @ivanovmg generally lgtm.
> 
> The removal of the setters in csvs is a great improvement, for readability and static typing.
> 
> I'm not sure about removing the TableFormatter class though. This provided a common base for LatexFormatter, HTMLFormatter and what is now StringFormatter, for code that does not belong in DataFrameFormatter.
> 
> I suggest that future refactors are more atomic. I see several themes here that probably should have been discussed/reviewed in isolation. The primary one being the adoption of DataFrameFormatter for use in csv formatting. I'm also not sure that DataFrameRenderer class is even necessary.

The main target here was to extract StringFormatter from DataFrameFormatter.
And I am glad I figured out how to separate it.

Regarding ``DataFrameRenderer``, we can probably think about it later. My idea was to keep it all in one place for all formats (currently not all).

My concern with TableFormatter was that some properties (attributes) are not related to some ancestors. For example, ``is_truncated`` has nothing to do with LatexFormatter. The same for ``formatters`` and presumably ``should_show_dimensions``.

simonjayhawkins commented:
> The main target here was to extract StringFormatter from DataFrameFormatter.
> And I am glad I figured out how to separate it.

agreed this is an important step for further refactoring. nice work.

> 
> Regarding `DataFrameRenderer`, we can probably think about it later. My idea was to keep it all in one place for all formats (currently not all).

yep. follow-ons ok.

> 
> My concern with TableFormatter was that some properties (attributes) are not related to some ancestors. For example, `is_truncated` has nothing to do with LatexFormatter. The same for `formatters` and presumably `should_show_dimensions`.

It was a bit difficult to get one's head around the Inheritance and Composition used together, and there was not a clear separation. I think that I anticipated, that over time the shared functionality in DataFrameFormatter would move to TableFormatter, DataFrameFormatter would evolve to be what is now StringFormatter and TableFormatter would evolve to be what is now DataFrameFormatter. So the latex, html and string formatters would have all inherited from a base class. But as you have noted, composition is also an option.

ivanovmg commented:
> 
> 
> > The main target here was to extract StringFormatter from DataFrameFormatter.
> > And I am glad I figured out how to separate it.
> 
> agreed this is an important step for further refactoring. nice work.
> 
> > Regarding `DataFrameRenderer`, we can probably think about it later. My idea was to keep it all in one place for all formats (currently not all).
> 
> yep. follow-ons ok.
> 
> > My concern with TableFormatter was that some properties (attributes) are not related to some ancestors. For example, `is_truncated` has nothing to do with LatexFormatter. The same for `formatters` and presumably `should_show_dimensions`.
> 
> It was a bit difficult to get one's head around the Inheritance and Composition used together, and there was not a clear separation. I think that I anticipated, that over time the shared functionality in DataFrameFormatter would move to TableFormatter, DataFrameFormatter would evolve to be what is now StringFormatter and TableFormatter would evolve to be what is now DataFrameFormatter. So the latex, html and string formatters would have all inherited from a base class. But as you have noted, composition is also an option.

In fact my first guess was to use inheritance. But then I realized that the ancestors will have only something in common with the parent TableFormatter. Thus, I figured out that the composition would be a better option.
Now, as the formatters are more or less separated, we can have a better look at what they have in common and probably improve the architecture (in a separate PR, once this is merged).

ivanovmg commented:
There are now failures related to parquet and datetime. Looks like these are persistent across multiple PRs today.

ivanovmg committed:
TYP: extract imports for typing only in csvs

ivanovmg committed:
TYP: move FloatFormatType alias to _typing

ivanovmg commented on pandas/io/formats/csvs.py:
```diff
@@ -29,19 +29,16 @@
 from pandas.core.indexes.api import Index
 
 from pandas.io.common import get_filepath_or_buffer, get_handle
+from pandas.io.formats.format import DataFrameFormatter, FloatFormatType
```
I implemented as you suggested.
General question: is it OK to have imports from ``pandas._typing`` on top, not inside ``if TYPE_CHECKING`` block?

simonjayhawkins replied:
yep. that's how we do it elsewhere. all imports in pandas._typing are inside a TYPE_CHECKING block so should be safe.

simonjayhawkins commented:
> Now, as the formatters are more or less separated, we can have a better look at what they have in common and probably improve the architecture (in a separate PR, once this is merged).

sgtm

jreback commented:
thanks @ivanovmg
